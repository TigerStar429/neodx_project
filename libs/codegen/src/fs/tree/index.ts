export { FsTree } from './impl/fs-tree';
export { ReadonlyVirtualFsTree } from './impl/readonly-virtual-fs-tree';
export { VirtualTree } from './impl/virtual-tree';
export type { FileChange, Tree } from './types';
export { FileChangeType } from './types';
