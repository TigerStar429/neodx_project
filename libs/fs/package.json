{
  "name": "@neodx/fs",
  "version": "0.0.1",
  "description": "File system helpers",
  "type": "module",
  "main": "dist/cjs/index.cjs",
  "types": "dist/types/index.d.ts",
  "module": "dist/mjs/index.mjs",
  "source": "src/index.ts",
  "sideEffects": false,
  "files": [
    "dist",
    "README.md",
    "package.json"
  ],
  "scripts": {
    "typecheck": "tsc --noEmit",
    "clean": "yarn dlx rimraf dist",
    "build": "libmake",
    "test": "NODE_OPTIONS='--experimental-vm-modules' jest",
    "lint": "eslint src"
  },
  "devDependencies": {
    "@neodx/config": "workspace:*",
    "@neodx/libmake": "workspace:*",
    "@types/jest": "29.2.5",
    "@types/node": "18.11.18",
    "@types/tmp": "^0.2.3",
    "eslint": "8.31.0",
    "eslint-kit": "6.10.0",
    "jest": "29.3.1",
    "tmp": "0.2.1",
    "typescript": "4.9.4"
  },
  "dependencies": {
    "@neodx/std": "workspace:*",
    "tiny-glob": "0.2.9"
  }
}
