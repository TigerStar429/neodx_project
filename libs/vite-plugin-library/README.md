# @neodx/builder

[Vite](https://vitejs.dev/) plugin for simple libraries setup

## Examples

### Basic usage

```typescript
import { defineConfig } from 'vite';
import library from '@neodx/vite-plugin-library';

export default defineConfig({
  plugins: [library()]
});
```

### Typescript

```shell
yarn add -D vite-tsconfig-paths vite-plugin-dts
```

```typescript
import { defineConfig } from 'vite';
import library from '@neodx/vite-plugin-library';
import tsconfigPaths from 'vite-tsconfig-paths';
import dts from 'vite-tsconfig-paths';

export default defineConfig({
  plugins: [tsconfigPaths(), dts(), library()]
});
```

### Generate package.json fields

```typescript
import { defineConfig } from 'vite';
import library from '@neodx/vite-plugin-library';
import tsconfigPaths from 'vite-tsconfig-paths';
import dts from 'vite-tsconfig-paths';

export default defineConfig({
  plugins: [
    tsconfigPaths(),
    dts(),
    library({
      updatePackageExports: true,
      updatePackageMainFields: true // defaults to "updatePackageExports"
    })
  ]
});
```

### Custom entries

#### One entry (other than "index.ts")

```typescript
export default defineConfig({
  plugins: [
    // ...
    library({
      entry: 'src/other.ts'
    })
  ]
});
```

#### Multiple static entries

```typescript
export default defineConfig({
  plugins: [
    // ...
    library({
      entry: ['src/index.ts', 'src/other/file.ts']
    })
  ]
});
```

#### Glob

```typescript
export default defineConfig({
  plugins: [
    // ...
    library({
      entry: [
        'src/index.ts',
        'src/*/index.ts',
        '!src/internal/index.ts',
        'src/features/*.ts',
        'src/features/*/index.ts'
      ]
    })
  ]
});
```
