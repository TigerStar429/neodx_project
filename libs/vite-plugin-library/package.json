{
  "name": "@neodx/vite-plugin-library",
  "version": "0.0.1",
  "description": "No description",
  "type": "module",
  "source": "src/index.ts",
  "main": "./dist/index.js",
  "types": "./dist/index.d.ts",
  "module": "./dist/index.js",
  "sideEffects": false,
  "files": [
    "dist",
    "README.md",
    "package.json"
  ],
  "scripts": {
    "typecheck": "tsc --noEmit",
    "build": "vite build",
    "build-alt": "libmake",
    "build-ts": "tsup src/index.ts --dts",
    "test": "vitest",
    "lint": "eslint src"
  },
  "dependencies": {
    "@neodx/codegen": "worskace:*",
    "@neodx/fs": "worskace:*",
    "@neodx/std": "worskace:*",
    "picocolors": "1.0.0",
    "pkg-types": "1.0.1"
  },
  "devDependencies": {
    "@types/node": "18.11.18",
    "eslint": "8.31.0",
    "eslint-kit": "6.10.0",
    "tsup": "6.5.0",
    "typescript": "4.9.4",
    "vite": "4.0.4",
    "vite-tsconfig-paths": "4.0.3"
  },
  "peerDependencies": {
    "vite": "4"
  }
}
