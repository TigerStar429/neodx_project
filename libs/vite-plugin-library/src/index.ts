export { vitePluginLibrary } from './vite-plugin-library';
