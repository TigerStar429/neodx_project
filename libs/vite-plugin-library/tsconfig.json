{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "types": ["node"]
  },
  "include": [".", "../*/src", "vite.config.ts"]
}
