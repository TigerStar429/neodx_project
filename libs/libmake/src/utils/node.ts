import type { Stats } from 'node:fs';
import { access, stat } from 'node:fs/promises';
import { dirname, resolve } from 'node:path';
import glob from 'tiny-glob';

export const isFile = (path: string) => assertPath(path, stats => stats.isFile());
export const isDirectory = (path: string) => assertPath(path, stats => stats.isDirectory());

export const rootDirname = (path: string): string => {
  const dir = dirname(path);

  return dirname(dir) === emptyDirname ? dir : rootDirname(dir);
};

export const exists = (path: string) =>
  access(path)
    .then(() => true)
    .catch(() => false);

export const assertPath = async (path: string, fn: (stats: Stats) => boolean) => {
  if (await exists(path)) {
    return fn(await stat(path));
  }
  return false;
};

export const findFiles = (patterns: string[], cwd: string) =>
  Promise.all(
    patterns.map(pattern =>
      isGlobPattern(pattern)
        ? glob(pattern, { cwd, absolute: false, filesOnly: true })
        : Promise.resolve(pattern)
    )
  ).then(results => results.flat().map(path => resolve(cwd, path)));

const isGlobPattern = (pattern: string) => pattern.includes('*');
const emptyDirname = dirname('');
