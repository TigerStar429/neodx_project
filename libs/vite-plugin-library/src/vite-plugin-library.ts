import type { Tree } from '@neodx/codegen';
import { ReadonlyVirtualFsTree, readTreeJson, updateTreeJson } from '@neodx/codegen';
import { scan } from '@neodx/fs';
import { compact, toArray } from '@neodx/std';
import { builtinModules } from 'node:module';
import type { PackageJson } from 'pkg-types';
import type { Plugin } from 'vite';
import { analyzeProject } from './analyze-project';
import { createExportsGenerator, ExportsGenerator } from './exports';

export interface VitePluginLibraryParams {
  entry?: string | string[];
  addTypes?: boolean;
  updatePackageExports?: boolean;
  updatePackageMainFields?: boolean;
}

export function vitePluginLibrary({
  entry: userEntry,
  addTypes,
  updatePackageExports,
  updatePackageMainFields = updatePackageExports
}: VitePluginLibraryParams = {}): Plugin {
  let tree: Tree;
  let exportsGenerator: ExportsGenerator | null = null;

  return {
    name: 'vite-plugin-library',
    async closeBundle() {
      if (exportsGenerator) {
        await updateTreeJson<PackageJson>(tree, 'package.json', prev => ({
          ...prev,
          ...exportsGenerator?.getFields(),
          exports: {
            ...(typeof prev.exports !== 'string' && prev.exports),
            ...exportsGenerator?.getExports()
          }
        }));
        console.log(await readTreeJson(tree, 'package.json'));
      }
    },
    generateBundle(options, files) {
      exportsGenerator?.addBundle(options, files);
    },
    async config({ root = process.cwd(), build }) {
      const { pkg, defaultEntry, deps, tsConfig } = await analyzeProject(root);
      const entryPatterns = compact(toArray(userEntry ?? defaultEntry));
      const entryFiles = await scan(root, entryPatterns);
      const entry = Object.fromEntries(
        entryFiles.map(file => [file.replace('src/', '').replace(/\.[tj]sx?$/, ''), file])
      );

      if (entryFiles.length === 0) {
        throw new Error(`Not found any entry file`);
      }
      tree = new ReadonlyVirtualFsTree(root);
      if (updatePackageExports || updatePackageMainFields) {
        exportsGenerator = createExportsGenerator({
          addTypes: addTypes ?? Boolean(tsConfig),
          outDir: build?.outDir,
          root
        });
      }
      return {
        build: {
          lib: {
            name: pkg.name,
            entry,
            formats: ['cjs', 'es']
          },
          rollupOptions: {
            external: [
              /node:.*/,
              // eslint-disable-next-line @typescript-eslint/require-array-sort-compare
              ...builtinModules.filter(x => !/^_|^(internal|v8|node-inspect)\/|\//.test(x)).sort(),
              ...deps.prod
            ]
          }
        }
      };
    }
  };
}
