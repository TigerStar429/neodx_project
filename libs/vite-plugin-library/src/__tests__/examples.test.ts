import { deepReadDir } from '@neodx/fs';
import { readdir } from 'node:fs/promises';
import { resolve } from 'node:path';
import { build } from 'vite';
import tsconfigPaths from 'vite-tsconfig-paths';
import { describe, expect, test } from 'vitest';
import { vitePluginLibrary, VitePluginLibraryParams } from '../vite-plugin-library';

const root = new URL(`../../`, import.meta.url).pathname;
const examplesRoot = resolve(root, 'examples');
const examplesOptions = {
  'multiple-ts-glob': {
    entry: ['src/index.ts', 'src/modules/*.ts', 'src/modules/*/index.ts']
  },
  'multiple-ts': {
    entry: ['src/index.ts', 'src/second-entry.ts'],
    addTypes: true,
    updatePackageExports: true
  }
} as Record<string, VitePluginLibraryParams>;

describe('examples', async () => {
  const examplesNames = await readdir(examplesRoot);

  test.each(examplesNames)(
    `build example "%s"`,
    async name => {
      const root = resolve(examplesRoot, name);
      const outDir = resolve(root, 'dist');

      await build({
        root,
        build: {
          outDir
        },
        plugins: [tsconfigPaths(), vitePluginLibrary(examplesOptions[name])]
      });
      expect(await deepReadDir(outDir)).toMatchSnapshot();
    },
    {
      timeout: 10_000
    }
  );
});
