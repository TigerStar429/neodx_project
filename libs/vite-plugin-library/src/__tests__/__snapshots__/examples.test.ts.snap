// Vitest Snapshot v1

exports[`examples > build example "advanced-mixed" 1`] = `
[
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/advanced-mixed/dist/index.cjs",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/advanced-mixed/dist/index.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/advanced-mixed/dist/style.css",
]
`;

exports[`examples > build example "basic-css" 1`] = `
[
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/basic-css/dist/dynamic-611e6833.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/basic-css/dist/dynamic-6ba6dc7f.mjs",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/basic-css/dist/index.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/basic-css/dist/index.mjs",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/basic-css/dist/style.css",
]
`;

exports[`examples > build example "basic-css-sass" 1`] = `
[
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/basic-css-sass/dist/index.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/basic-css-sass/dist/index.mjs",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/basic-css-sass/dist/style.css",
]
`;

exports[`examples > build example "basic-ts" 1`] = `
[
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/basic-ts/dist/index.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/basic-ts/dist/index.mjs",
]
`;

exports[`examples > build example "feat-ts-without-definitions" 1`] = `
[
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/feat-ts-without-definitions/dist/index.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/feat-ts-without-definitions/dist/index.mjs",
]
`;

exports[`examples > build example "multiple-ts" 1`] = `
[
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts/dist/index.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts/dist/index.mjs",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts/dist/second-entry.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts/dist/second-entry.mjs",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts/dist/shared-39372394.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts/dist/shared-f0141ee8.mjs",
]
`;

exports[`examples > build example "multiple-ts-glob" 1`] = `
[
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts-glob/dist/index-930f9e54.mjs",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts-glob/dist/index-9f0d2d81.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts-glob/dist/index.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts-glob/dist/index.mjs",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts-glob/dist/modules/bar/index.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts-glob/dist/modules/bar/index.mjs",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts-glob/dist/modules/baz.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts-glob/dist/modules/baz.mjs",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts-glob/dist/modules/foo/index.js",
  "/Users/primary/WebstormProjects/neodx/libs/vite-plugin-library/examples/multiple-ts-glob/dist/modules/foo/index.mjs",
]
`;
